src/auth.ts

```diff
@@ -1,6 +1,33 @@
 import { User } from './config_types';
 
-export interface Authenticator {
-    isAuthProvider(user: User): boolean;
-    getToken(user: User): string | null;
+interface CredentialsBase {
+    expirationTimestamp?: string;
+}
+
+export interface TokenCredentials extends CredentialsBase {
+    type: 'token';
+    token: string;
+}
+
+export interface ClientCertCredentials extends CredentialsBase {
```

Comment:
This logic feels a bit leaky. 

Later you do checks `if type == 'token'` I feel like adding these extra types here complicates the code I think it would be more clear if this was in a single type or we should figure out how to abstract the logic to provide unified function (which I don't think is possible).

This fits more in line with the rest of the codebase (config parsing) and the go-client.

Reply:
I think singular here makes more sense too.

Reply:
I agree, I think that the right thing is to have `Credential.apply(opts)` instead of leaking the credential into the authentication flow.